visitors/http_rest/spec_visitor.go

```diff
@@ -514,69 +515,109 @@ func extendContext(cin Context, node interface{}) {
 		}
 	case *pb.Data:
 		// Update the context for the field about to be visited.
-		// HTTPMeta is only valid for the top-level Data object.
-		if node.GetMeta() != nil && node.GetMeta().GetHttp() != nil {
-			ctx.setTopLevelDataIndex(len(ctx.GetRestPath()) - 1)
-
-			// Figure out whether the request or response is being visited, and set
-			// the response code.
-			meta := node.GetMeta().GetHttp()
-			switch rc := meta.GetResponseCode(); rc {
-			case 0: // arg
-				ctx.setIsArg(true)
-				ctx.appendRestPath("Arg")
-			default:
-				ctx.setIsArg(false)
-				ctx.appendRestPath("Response")
-
-				responseCode := "default"
-				if rc != -1 {
-					responseCode = strconv.Itoa(int(rc))
+		//
+		// HTTPMeta is only populated for the top-level Data objects in a request or
+		// response, and for the top-level Data object in each part of a multi-part
+		// body.
+		if meta := node.GetMeta().GetHttp(); meta != nil {
+			// Multipart bodies have two layers of metadata: an outer layer associated
+			// with the Data node representing the multiple parts, and an inner layer
+			// at the Data nodes representing each individual part. Determine whether
+			// we are at the inner layer, and if so, try to get the name of the
+			// current body part.
+			//
+			// ugh.
+			isMultipartBodyMember := false
+			multipartBodyMemberNameOpt := optionals.None[string]()
+			if meta.GetBody() != nil {
+				outerData, _ := ctx.GetInnermostNode(reflect.TypeOf((*pb.Data)(nil)))
+				if outerData != nil {
+					isMultipartBodyMember = outerData.(*pb.Data).GetMeta().GetHttp().GetMultipart() != nil
+					if isMultipartBodyMember {
+						// Get the name of the multipart body.
+						astPath := ctx.GetPath()
+						if !astPath.IsEmpty() {
+							astPathEdge := astPath.GetLast().OutEdge
+							if mapValueEdge, ok := astPathEdge.(*MapValueEdge); ok {
+								multipartBodyMemberNameOpt = optionals.Some(fmt.Sprint(mapValueEdge.MapKey))
+							}
+						}
+					}
+				}
+			}
+
+			if !isMultipartBodyMember {
+				ctx.setTopLevelDataIndex(len(ctx.GetRestPath()) - 1)
+
+				// Figure out whether the request or response is being visited, and set
+				// the response code.
+				switch rc := meta.GetResponseCode(); rc {
+				case 0: // arg
+					ctx.setIsArg(true)
+					ctx.appendRestPath("Arg")
+				default:
+					ctx.setIsArg(false)
+					ctx.appendRestPath("Response")
+
+					responseCode := "default"
+					if rc != -1 {
+						responseCode = strconv.Itoa(int(rc))
+					}
+					ctx.appendRestPath(responseCode)
+					ctx.setResponseCode(responseCode)
 				}
-				ctx.appendRestPath(responseCode)
-				ctx.setResponseCode(responseCode)
 			}
 
 			// Figure out the name and kind of parameter being visited. If visiting a
 			// body, also figure out its content type.
-			var name string
-			named := true
+			restPathNames := []string{}
+			fieldPathName := optionals.None[string]()
```

Comment:
I seem to recall the context path was an allocation hot-spot, so if this array allocation could be avoided that would be desirable, but I don't see a good way to do it other than move RestPath manipulation into each of the cases.